### PR TITLE
Add table aws_glacier_vault closes #149

### LINKED ISSUE #149: Add table aws_glacier_vault
**References**
https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.37.24/service/glacier

### PR DESCRIPTION
**Pull request by Paulami30**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_glacier_vault []

PRETEST: tests/aws_glacier_vault

TEST: tests/aws_glacier_vault
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_glacier_vault.named_test_resource: Creating...
aws_glacier_vault.named_test_resource: Still creating... [10s elapsed]
aws_glacier_vault.named_test_resource: Creation complete after 16s [id=turbottest88038]
aws_glacier_vault_lock.vault_policy: Creating...
aws_glacier_vault_lock.vault_policy: Creation complete after 3s [id=turbottest88038]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = arn:aws:glacier:us-east-1:352685002396:vaults/turbottest88038
resource_id = turbottest88038
resource_name = turbottest88038

Running SQL query: test-get-query.sql
[
  {
    "tags_src": {
      "name": "turbottest88038"
    },
    "title": "turbottest88038",
    "vault_name": "turbottest88038"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest88038"
    ],
    "policy": {
      "Statement": [
        {
          "Action": [
            "glacier:InitiateJob",
            "glacier:GetJobOutput"
          ],
          "Effect": "Allow",
          "Principal": "*",
          "Resource": "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest88038",
          "Sid": "__default_policy_ID"
        }
      ],
      "Version": "2012-10-17"
    },
    "tags": {
      "name": "turbottest88038"
    },
    "title": "turbottest88038",
    "vault_lock_policy_std": {
      "Statement": [
        {
          "Action": [
            "glacier:deletearchive"
          ],
          "Condition": {
            "NumericLessThan": {
              "glacier:archiveageindays": [
                "365"
              ]
            }
          },
          "Effect": "Deny",
          "Principal": {
            "AWS": [
              "*"
            ]
          },
          "Resource": [
            "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest88038"
          ],
          "Sid": "deny-based-on-archive-age"
        }
      ],
      "Version": "2012-10-17"
    },
    "vault_name": "turbottest88038"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "akas": [
      "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest88038"
    ],
    "tags": {
      "name": "turbottest88038"
    },
    "title": "turbottest88038",
    "vault_name": "turbottest88038"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

POSTTEST: tests/aws_glacier_vault

TEARDOWN: tests/aws_glacier_vault

SUMMARY:

1/1 passed.

```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```

> select
  vault_name,
  creation_date,
  last_inventory_date,
  number_of_archives,
  size_in_bytes
from
  aws_glacier_vault;
+------------+---------------------+---------------------+--------------------+---------------+
| vault_name | creation_date       | last_inventory_date | number_of_archives | size_in_bytes |
+------------+---------------------+---------------------+--------------------+---------------+
| test001    | 2021-03-31 07:12:16 | <null>              | 0                  | 0             |
| test       | 2021-03-30 07:10:04 | <null>              | 0                  | 0             |
+------------+---------------------+---------------------+--------------------+---------------+

> select
  title,
  p as principal,
  a as action,
  s ->> 'Effect' as effect,
  s -> 'Condition' as conditions
from
  aws_glacier_vault,
  jsonb_array_elements(policy_std -> 'Statement') as s,
  jsonb_array_elements_text(s -> 'Principal' -> 'AWS') as p,
  jsonb_array_elements_text(s -> 'Action') as a
where
  s ->> 'Effect' = 'Allow'
  and a in ('*', 'glacier:*');
+-------+-----------+-----------+--------+------------+
| title | principal | action    | effect | conditions |
+-------+-----------+-----------+--------+------------+
| test  | *         | glacier:* | Allow  | <null>     |
+-------+-----------+-----------+--------+------------+                           

> select
  title,
  p as principal,
  a as action,
  s ->> 'Effect' as effect,
  s -> 'Condition' as conditions
from
  aws_glacier_vault,
  jsonb_array_elements(policy_std -> 'Statement') as s,
  jsonb_array_elements_text(s -> 'Principal' -> 'AWS') as p,
  jsonb_array_elements_text(s -> 'Action') as a
where
  p = '*'
  and s ->> 'Effect' = 'Allow';
+-------+-----------+-----------+--------+------------+
| title | principal | action    | effect | conditions |
+-------+-----------+-----------+--------+------------+
| test  | *         | glacier:* | Allow  | <null>     |
+-------+-----------+-----------+--------+------------+

> select
  title,
  a as action,
  s ->> 'Effect' as effect,
  s -> 'Condition' -> 'NumericLessThan' ->> 'glacier:archiveageindays' as archive_age_in_days
from
  aws_glacier_vault,
  jsonb_array_elements(vault_lock_policy_std -> 'Statement') as s,
  jsonb_array_elements_text(s -> 'Action') as a;
+---------+-----------------------+--------+---------------------+
| title   | action                | effect | archive_age_in_days |
+---------+-----------------------+--------+---------------------+
| test001 | glacier:deletearchive | Deny   | ["365"]             |
| test    | glacier:*             | Allow  | <null>              |
+---------+-----------------------+--------+---------------------+

> select
  vault_name,
  tags
from
  aws_glacier_vault
where
  not tags :: JSONB ? 'owner';
+------------+------+
| vault_name | tags |
+------------+------+
| test       | {}   |
| test001    | {}   |
+------------+------+

```
</details>

### CODE REVIEW

File: aws/table_aws_glacier_vault.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+//// TABLE DEFINITION
+
+func tableAwsGlacierVault(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_glacier_vault",
+		Description: "AWS Glacier Vault",
+		Get: &plugin.GetConfig{
+			KeyColumns:        plugin.SingleColumn("vault_name"),
+			ShouldIgnoreError: isNotFoundError([]string{"ResourceNotFoundException"}),
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:22]: Handle below error
```
Error: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, DescribeVaultInput.VaultName.
```

File: aws/table_aws_glacier_vault.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+			{
+				Name:        "policy_std",
+				Description: "Contains the policy in a canonical form for easier searching.",
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getGlacierVaultAccessPolicy,
+				Transform:   transform.FromField("Policy.Policy").Transform(unescape).Transform(policyToCanonical),
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags associated with the cluster.",
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:77]:
```suggestion
				Description: "A list of tags associated with the vault.",
```

File: aws/table_aws_glacier_vault.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+	vaultName := h.Item.(*glacier.DescribeVaultOutput)
+
+	commonData, err := getCommonColumns(ctx, d, h)
+	if err != nil {
+		return nil, err
+	}
+
+	commonColumnData := commonData.(*awsCommonColumnData)
+	accountID := commonColumnData.AccountId
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:205]: Can't we use the `VaultARN` to get the accountID?

File: aws/table_aws_glacier_vault.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:245]: Can't we use the `VaultARN` to get the accountID?

File: aws/table_aws_glacier_vault.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+				Description: resourceInterfaceDescription("title"),
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("VaultName"),
+			},
+			{
+				Name:        "tags",
+				Description: resourceInterfaceDescription("tags"),
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     listTagsForGlacierVault,
+				Transform:   transform.FromField("Tags"),
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:95]:
```suggestion
```

File: aws/table_aws_glacier_vault.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	// Build param
+	param := &glacier.ListTagsForVaultInput{
+		VaultName: vaultName.VaultName,
+		AccountId: aws.String(accountID),
+	}
+
+	vaultTags, err := svc.ListTagsForVault(param)
+
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:260]:
```suggestion
```

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
Make suggested changes

**Comment by Paulami30** [comment]
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_glacier_vault []

PRETEST: tests/aws_glacier_vault

TEST: tests/aws_glacier_vault
Running terraform
data.aws_region.alternate: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
data.aws_partition.current: Refreshing state...
data.aws_region.primary: Refreshing state...
data.null_data_source.resource: Refreshing state...
aws_glacier_vault.named_test_resource: Creating...
aws_glacier_vault.named_test_resource: Still creating... [10s elapsed]
aws_glacier_vault.named_test_resource: Creation complete after 10s [id=turbottest99510]
data.template_file.resource_aka: Refreshing state...

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

resource_aka = arn:aws:glacier:us-east-1:352685002396:vaults/turbottest99510
resource_id = turbottest99510
resource_name = turbottest99510

Running SQL query: test-get-query.sql
[
  {
    "tags_src": {
      "name": "turbottest99510"
    },
    "title": "turbottest99510",
    "vault_name": "turbottest99510"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "akas": [
      "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest99510"
    ],
    "policy": {
      "Statement": [
        {
          "Action": [
            "glacier:InitiateJob",
            "glacier:GetJobOutput"
          ],
          "Effect": "Allow",
          "Principal": "*",
          "Resource": "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest99510",
          "Sid": "__default_policy_ID"
        }
      ],
      "Version": "2012-10-17"
    },
    "tags": {
      "name": "turbottest99510"
    },
    "title": "turbottest99510",
    "vault_name": "turbottest99510"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "akas": [
      "arn:aws:glacier:us-east-1:352685002396:vaults/turbottest99510"
    ],
    "tags": {
      "name": "turbottest99510"
    },
    "title": "turbottest99510",
    "vault_name": "turbottest99510"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

POSTTEST: tests/aws_glacier_vault

TEARDOWN: tests/aws_glacier_vault

SUMMARY:

1/1 passed.

File: aws/table_aws_glacier_vault.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	// TODO put me in helper function
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+
+	vaultName := h.Item.(*glacier.DescribeVaultOutput)
+	splitID := strings.Split(string(*vaultName.VaultARN), ":")
+	accountID := splitID[4]
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:199]:
```suggestion
	accountID := strings.Split(*vaultName.VaultARN, ":")[4]
```

File: aws/table_aws_glacier_vault.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	logger.Trace("getGlacierVaultAccessPolicy")
+
+	// TODO put me in helper function
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+
+	vaultName := h.Item.(*glacier.DescribeVaultOutput)
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:197]:
```suggestion
	data := h.Item.(*glacier.DescribeVaultOutput)
```

File: aws/table_aws_glacier_vault.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+	splitID := strings.Split(string(*vaultName.VaultARN), ":")
+	accountID := splitID[4]
```
Comment (Subhajit97) [inline aws/table_aws_glacier_vault.go:233]:
```suggestion
	data := h.Item.(*glacier.DescribeVaultOutput)
	accountID := strings.Split(*data.VaultARN, ":")[4]
```

**Review by Subhajit97 — DISMISSED** [review]

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
All example looks good, but if we can add the archival days or archival policy of the vault will be good, this will help to list the buckets nearing the last date of deletion kind of.

**Comment by rajmohanty17** [comment]
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Glacier.html#describeVault-property

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**